server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2999,14 +2998,14 @@ public void onResponse(final Releasable releasable) {
                 } catch (final Exception e) {
                     if (combineWithAction == null) {
                         // otherwise leave it to combineWithAction to release the permit
-                        releaseOnce.run();
+                        releaseOnce.close();
                     }
                     innerFail(e);
                 } finally {
                     if (combineWithAction != null) {
                         combineWithAction.onResponse(releasable);
                     } else {
-                        releaseOnce.run();
+                        releaseOnce.close();
```

Comment:
Won't this have already been run above in the catch?

Reply:
Right, the close in the catch seems redundant. We do however close first and then fail the engine there after. Maybe that matters in some cases. I'll look into that separately, seems like the kind of thing to be very careful about :)